rotkehlchen/exchanges/kucoin.py

```diff
@@ -0,0 +1,775 @@
+            # NB: id only exists for withdrawals
+            link_id = raw_result['id'] if category_case == 'withdrawal' else transaction_id
+        except (DeserializationError, KeyError, ValueError) as e:
+            msg = str(e)
+            if isinstance(e, KeyError):
+                msg = f'Missing key in {category_case}: {msg}.'
+
+            log.error(
+                f'Failed to deserialize a kucoin {category_case}',
+                error=msg,
+                raw_result=raw_result,
+            )
+            raise DeserializationError(f'Failed to deserialize a kucoin {category_case}.') from e
+
+        try:
+            fee_asset = asset_from_kucoin(fee_currency_symbol)
+        except DeserializationError as e:
+            log.error(
+                f'Unexpected asset symbol type in a kucoin {category_case}',
+                error=str(e),
+                raw_result=raw_result,
+            )
+            raise DeserializationError(f'Failed to deserialize a kucoin {category_case}.') from e
+        except (UnknownAsset, UnsupportedAsset) as e:
+            asset_tag = 'unknown' if isinstance(e, UnknownAsset) else 'unsupported'
+            raise DeserializationError(
+                f'Found {asset_tag} kucoin asset {e.asset_name} while processing a {category_case}. '  # noqa: E501
+                f'Ignoring it.',
+            ) from e
+
+        asset_movement = AssetMovement(
+            timestamp=timestamp,
+            location=Location.KUCOIN,
+            category=category,
+            address=address,
+            transaction_id=transaction_id,
+            asset=fee_asset,
+            amount=amount,
+            fee_asset=fee_asset,
+            fee=fee,
+            link=link_id,
+        )
+        return asset_movement, None
+
+    @staticmethod
+    def _deserialize_trade(
+            raw_result: Dict[str, Any],
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> Tuple[Optional[Trade], Optional[SkipReason]]:
+        """Process a trade result and deserialize it
+
+        May raise DeserializationError
+        """
+        try:
+            timestamp = deserialize_timestamp(int(raw_result['createdAt'] / 1000))
+            if timestamp > end_ts:
+                return None, SkipReason.ABOVE_TIMESTAMP_RANGE
+            if timestamp < start_ts:
+                return None, SkipReason.BELOW_TIMESTAMP_RANGE
+
+            trade_type = TradeType.BUY if raw_result['side'] == 'buy' else TradeType.SELL
+            amount = deserialize_asset_amount(raw_result['size'])
+            rate = deserialize_price(raw_result['price'])
+            fee = deserialize_fee(raw_result['fee'])
+            trade_id = raw_result['tradeId']
+            fee_currency_symbol = raw_result['feeCurrency']
+            base_asset_symbol, quote_asset_symbol = raw_result['symbol'].split('-')
+        except (DeserializationError, KeyError, ValueError) as e:
```

Comment:
I've been following your recommendation of logging as close as possible to the error (e.g. inside the deserialization method) and trying to make DeserializationError the only exception that a deserialization method can raise. I've also tried to standardise the dev and user messages (example below):
```python
  log.error(
      'Failed to deserialize a kucoin trade',  // standardised
      error=msg,  // per exception
      raw_result=raw_result,
  )
  raise DeserializationError('Failed to deserialize a kucoin trade.')  // standardised, don't show the error problem but just the context (i.e. the exchange and the case) unless it comes from an unsupported token, as it has to be more user friendly and contain the asset name
```

If I don't catch the DeserializationError comming from any of `deserialize_timestamp, deserialize_asset_amount, deserialize_price and deserialize_fee`, the `_api_query_paginated` I will need to add there another `log.error` for including the `raw_result` no? Besides as these exceptions won't have a message that say `Failed to deserialize a kucoin whatever` they will be out of context and displaying the real error (e.g. invalid timestamp None) in the front-end. It would also display twice the text "Failed to deserialize..." in some cases:
```python
  log.error(
      'Failed to deserialize a kucoin trade',
      error=str(e),
      raw_result=raw_result,
  )
  self.msg_aggregator.add_error(
      f'Failed to deserialize a kucoin {case}. {str(e)}. Check logs for more details',  // "Failed to deserialize..." can appear twice if the exception message already has this sentence.
  )
  continue
```

I can move the `log.error` (with the raw_entry) to `_api_query_paginated` and change the exception messages of the deserialization methods. Let me know.

Reply:
I see the problem. Perhaps the good long term solution is to add a "location" argument to all these deserialize functions so the error message has more information when raised from there and no need to add them when processing them while they bubble up.

But for now perhaps keep it as it, and change the re-raised DeserializationError to a RemoteError. I think this is how we do it in some other places.

Reply:
You mean raise RemoteError inside `_deserialize_trade` and `_deserialize_asset_movement`? (swap all the DeserializationError for RemoteError). I haven't seen it in other exchanges.

Reply:
Hmmm no scratch that. I just don't like that you have it both here and in paginated. Perhaps what you say here is best:

> I can move the log.error (with the raw_entry) to _api_query_paginated and change the exception messages of the deserialization methods. Let me know.

Reply:
Ok, I'll do!